raiden-ts/src/channels/epics/settle.ts

```diff
@@ -66,56 +67,56 @@ export function channelSettleableEpic(
- * settleable, to give time for confirmation and for partner to settle.
+ * If partner is a LC and not the closing side, they're expected to wait [config.revealTimeout]
+ * after channel becomes settleable before attempting to auto-settle, so we should attempt it
+ * earlier [config.confirmationBlocks] after settleable. PCs always attempt earlier, so we should
+ * wait longer regardless of who closed the channel, to avoid races and wasted gas; if even after
+ * waiting (earlier or later) channel isn't settling/settled, we do it anyway.
  *
  * @param action$ - Observable of RaidenActions
  * @param state$ - Observable of RaidenStates
  * @param deps - RaidenEpicDeps members
  * @param deps.config$ - Config observable
+ * @param deps.address - Our address
  * @returns Observable of channelSettle.request actions
  */
 export function channelAutoSettleEpic(
-  {}: Observable<RaidenAction>,
+  action$: Observable<RaidenAction>,
   state$: Observable<RaidenState>,
-  { config$ }: RaidenEpicDeps,
+  { address, config$ }: RaidenEpicDeps,
 ): Observable<channelSettle.request> {
   return state$.pipe(
-    groupChannel(),
-    mergeMap((grouped$) =>
-      grouped$.pipe(
-        filter(
-          (channel): channel is Channel & { state: ChannelState.settleable } =>
-            channel.state === ChannelState.settleable,
-        ),
-        take(1),
-        withLatestFrom(config$),
-        // wait [confirmationBlocks, 2 * confirmationBlocks] before proceeding
-        delayWhen(([channel, { confirmationBlocks }]) => {
-          const settleBlock =
-            channel.closeBlock +
-            channel.settleTimeout +
-            Math.round(confirmationBlocks * (1.0 + Math.random()));
-          return state$.pipe(
-            pluck('blockNumber'),
-            filter((blockNumber) => settleBlock <= blockNumber),
-            take(1),
-          );
-        }),
-        withLatestFrom(state$),
-        // filter channel isn't yet being settled by us or partner (state=settling)
-        filter(
-          ([[channel], state]) =>
-            state.channels[channelKey(channel)]?.state === ChannelState.settleable,
-        ),
-        map(([[channel]]) =>
-          channelSettle.request(undefined, {
-            tokenNetwork: channel.tokenNetwork,
-            partner: channel.partner.address,
-          }),
-        ),
-      ),
-    ),
+    distinctUntilKeyChanged('blockNumber'),
+    withLatestFrom(action$.pipe(getPresencesByAddress()), config$),
+    mergeMap(function* ([
+      { blockNumber, channels },
+      presences,
+      { confirmationBlocks, revealTimeout },
+    ]) {
+      for (const channel of Object.values(channels)) {
+        if (channel.state !== ChannelState.settleable) continue;
+
+        const partnerIsOnlineLC =
+          channel.partner.address in presences &&
+          !getCap(presences[channel.partner.address]?.payload.caps, Capabilities.DELIVERY);
```

Comment:
Do we have such checks (having delivery cap on signals python client) in other places? If so, we should factor them out into one place.

Reply:
Yes, I think in one other place. I'll make a new commit to split this out in a separate function.

Reply:
:heavy_check_mark: